common/lib/xmodule/xmodule/library_content_module.py

```diff
@@ -371,13 +388,186 @@ def get_child_descriptors(self):
+            'invalid' (set) of dropped (block_type, block_id) tuples that are no longer valid
+            'overlimit' (set) of dropped (block_type, block_id) tuples that were previously selected
+            'added' (set) of newly added (block_type, block_id) tuples
+
+        Note that this implementation ignores the `max_count` argument:
+        The list of `selected` (block_type, block_id) tuples assigned to this student
+        has been determined based on information from external service providing adaptive learning features,
+        so we only remove blocks that are no longer valid here.
+        """
+        selected = set(tuple(k) for k in selected)  # set of (block_type, block_id) tuples assigned to this student
+
+        # Determine which of our children we will show:
+        valid_block_keys = set([(c.block_type, c.block_id) for c in children])
+        # Remove any selected blocks that are no longer valid:
+        invalid_block_keys = (selected - valid_block_keys)
+        if invalid_block_keys:
+            selected -= invalid_block_keys
+
+        # `overlimit_block_keys`:
+        # External service determines number of children to show,
+        # so no need to *drop* any blocks based on value of `max_count` argument.
+
+        # `added_block_keys`:
+        # External service determines number of children to show,
+        # so no need to *add* any blocks based on value of `max_count` argument.
+
+        return {
+            'selected': selected,
+            'invalid': invalid_block_keys,
+            # External service determines number of children to show,
+            # so `max_count` setting has no effect:
+            # We never *drop* previously selected children because of it,
+            # so the set of dropped children is empty by definition:
+            'overlimit': set(),
+            # External service determines number of children to show,
+            # so `max_count` setting has no effect:
+            # We never *add* any children because of it,
+            # so the set of added children is empty by definition:
+            'added': set(),
+        }
+
+    def selected_children(self):
+        """
+        Returns a set() of block_ids indicating which of the possible children
+        have been selected to display to the current user.
+        """
+        if hasattr(self, "_selected_set"):
+            # Already done:
+            return self._selected_set  # pylint: disable=access-member-before-definition
+
+        # Determine children to display based on information about pending reviews from external service.
+        # This needs to happen here (at instance level) because we need access to course and block-specific data,
+        # which is not available at the class level.
+        selected = self.get_selections_current_user(self.children)
+        block_keys = self.make_selection(selected, self.children, self.max_count, "adaptive")  # pylint: disable=no-member
+
```

Comment:
@pomegranited I don't think there is a nice way to DRY out the `selected_children` methods (at least not without doing a larger refactoring). For example, what you're suggesting won't work because `self.selected` needs to be updated *after* publishing analytics events: One of the methods that handle publishing events uses `self.selected` to compute information about state as it exists prior to updating that field.

It would be possible to introduce some decorators (e.g. for handling the `hasattr` check). But the selection logic is the most complicated part of this code, and I think we'd be trading readability/understandability for DRYness. Since the `selected_children` methods are also pretty short I think it makes sense to leave them as is for now.